File: Sprint4/J/C++/main.cpp
// https://contest.yandex.ru/contest/23991/problems/J/
#include <iostream>
#include <vector>
#include <algorithm>
#include <map>
#include <set>

int main() {
    int n;
    long long s;
    std::cin >> n;
    std::cin >> s;

    std::vector<int> arr(n);
    for (int i = 0; i < n; i++)
        std::cin >> arr[i];

    std::map<long long, std::vector<std::pair<long long, long long >>> hash_pairs;

    for (int i = 0; i < n - 1; i++) {
        for (int j = i + 1; j < n; j++) {
            hash_pairs[arr[i] + arr[j]].emplace_back(i, j);
        }
    }

    std::set<std::vector<int>> result;

    for (int i = 0; i < n - 1; i++) {
        for (int j = i + 1; j < n; j++) {
            long long sum = arr[i] + arr[j];
            if (hash_pairs.find(s - sum) != hash_pairs.end()) {
                for (auto &p: hash_pairs[s - sum]) {

                    if (p.first != i && p.first != j && p.second != i && p.second != j) {
                        std::vector<int> quad = {arr[i], arr[j], arr[p.first], arr[p.second]};
                        sort(quad.begin(), quad.end());
                        result.insert(quad);
                    }
                }
            }
        }
    }

    std::cout << result.size() << std::endl;
    for (auto &quad: result) {
        for (int num: quad)
            std::cout << num << " ";
        std::cout << std::endl;
    }

    return 0;
}

File: Sprint4/J/Rust/src/main.rs
// https://contest.yandex.ru/contest/23991/problems/J/

use std::collections::{BTreeSet, HashMap};

fn main() {
    let mut input = String::new();
    std::io::stdin().read_line(&mut input).unwrap();
    let n: usize = input.trim().parse().unwrap();

    let mut input = String::new();
    std::io::stdin().read_line(&mut input).unwrap();
    let s: i64 = input.trim().parse().unwrap();

    let mut arr = Vec::with_capacity(n);
    input.clear();
    std::io::stdin().read_line(&mut input).unwrap();
    arr.extend(
        input
            .trim()
            .split_whitespace()
            .map(|s| s.parse::<i64>().unwrap()),
    );

    let mut hash_pairs: HashMap<i64, Vec<(usize, usize)>> = HashMap::new();

    for i in 0..n - 1 {
        for j in i + 1..n {
            let sum = arr[i] + arr[j];
            hash_pairs.entry(sum).or_insert_with(Vec::new).push((i, j));
        }
    }

    let mut result: BTreeSet<Vec<i64>> = BTreeSet::new();

    for i in 0..n - 1 {
        for j in i + 1..n {
            let sum = arr[i] + arr[j];
            if let Some(pairs) = hash_pairs.get(&(s - sum)) {
                for &(first, second) in pairs {
                    if first != i && first != j && second != i && second != j {
                        let mut quad = vec![arr[i], arr[j], arr[first], arr[second]];
                        quad.sort_unstable();
                        result.insert(quad);
                    }
                }
            }
        }
    }

    println!("{}", result.len());
    for quad in result {
        for num in quad {
            print!("{} ", num);
        }
        println!();
    }
}
